### ISSUE TITLE
Support ngc to compile single file

### ISSUE THREAD

**Issue by xiaoxiangmoe** [angular/angular#37585]
# 🚀 feature request

### Relevant Package

 This feature request is for @angular/compiler and @angular/compiler-cli?

### Description

TypeScript has [transpileModule](https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API#a-simple-transform-function) to compile single file to js. Can we also support ngc to compile single file? Can we have a similar mechanism? It will help us to write plugins for snowpack and vite.

**Comment by JoostK** [comment]
The current characteristics of Angular do not allow for this mode of operation, as AOT compilations typically depend on other source files for local metadata (e.g. what directives are available to a component as determined by the NgModule class that declares the component) and dependency metadata in `.d.ts` files for extracting relevant information from libraries. Therefore, a full program is required.

**Comment by pauldraper** [comment]
@JoostK there appears to be some misunderstandings about how the Ivy compiler works.

> Currently Angular relies on global compilation...The compiler should use only information provided by **component decorator and its class and nothing else.**
>
> https://dev.to/eugeniolentini/angular-ivy-a-detailed-introduction-oj1#locality

> Ivy follows the locality principle where **only one file is compiled at a time.**
>
> https://medium.com/js-imaginea/ivy-a-look-at-the-new-render-engine-for-angular-953bf3b4907a

(Do you have any better resources about how it actually works and how to use it?)

**Comment by JoostK** [comment]
@pauldraper the definition of "locality" has been poorly understood and communication around this subject hasn't been the most clear. Over time, it has become clear that locality is not simply a binary concept, but rather a sliding scale.

ViewEngine employed truly global compilation, where the application was compiled together with all libraries. An artifact of this was that library components could be compiled with the knowledge of how it was being used in an application, making it truly global. This approach prevents independent compilation of the library ahead of time.

In Ivy, the scale has shifted a bit such that libraries can be compiled independent from a consuming application. Within the library compilation unit all information for compiling components is available. It does however still require an analysis step across the whole compilation unit, as components must know their declaring NgModule and its declarations, as references to used directives/components/pipes have to be emitted into the compiled template code. In the team we have referred to this as "compilation scope", unlike ViewEngine's "global scope".

Even if the concept of NgModule's was revised such that component compilation would not depend on them, it would still not allow for single file compilation. The reason is that the Angular compiler may need to statically evaluate certain pieces of information across files, e.g. when a selector is imported:

```ts
import {Directive} from '@angular/core';
import {SELECTOR} from './selectors';

@Directive({
  selector: SELECTOR,
})
export class ExampleDir {}
```

Additionally, the compiler needs the `.d.ts` files from libraries to know that e.g. `NgIf` has a selector of `[ngIf]`.

**Comment by pauldraper** [comment]
TypeScript has taken great pains to ensure that the language permits per-file compilation (there are rare exceptions like const enums).

It's disappointing that as a "wrapper of tsc", ngc hasn't abided by the same design choices.

---

I'm going to copy my closed-as-duplicate #39678:

> Both TypeScript and React (TSX) can compile much, much faster (like 10x) with ts.transpileModule.
>
> This just transpiles, without loading other modules or running type verification.
>
> AFAIK, no similar feature exists for Angular.

Slightly different than this original request for this issue, but I am hoping that even if per-file compilation is impossible, there could at least be some sort of "fast mode," like that operates on an AST but without TS type verification needing to be run.

**Comment by alxhub** [comment]
To reiterate, it's not a design choice of `ngc` but rather the fundamental design of Angular itself that precludes per-file compilation. Angular components don't exist in isolation - they may have external templates, for example. They also participate in the NgModule graph, which defines the set of dependencies (other components, directives, pipes) they're able to reference.

`transpileModule` is simply incompatible with the current design of Angular - there's no real way for us to provide it as an API at this time.

Additionally, `ngc` does make limited use of TypeScript's type information in order to process Angular decorators, since we must effectively "execute" JS syntax in order to statically understand decorator declarations. So a type-less compilation is not possible.

However, it's worth noting that the concept of a single-file transpilation _is_ the foundation of the [Angular Linker](https://github.com/angular/angular/issues/38366) design. After passing through `ngc` in its partial compilation mode, a given library's JS code can be transpiled for use with a specific version of Angular on a file-by-file basis.
